The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11293816473/job/31412850290) [(plain)](https://github.com/rust-lang/rust/commit/ab859f15c3f9410c9c4fad8ce068002b1f658b70/checks/31412850290/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 3.036 Building wheels for collected packages: reuse
#13 3.037   Building wheel for reuse (pyproject.toml): started
#13 3.287   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.288   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 3.288   Stored in directory: /tmp/pip-ephem-wheel-cache-2m2zeqnm/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.291 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.679 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.680 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.199 Collecting virtualenv
#13 4.199 Collecting virtualenv
#13 4.237   Downloading virtualenv-20.26.6-py3-none-any.whl (6.0 MB)
#13 4.320      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.0/6.0 MB 74.4 MB/s eta 0:00:00
#13 4.376 Collecting platformdirs<5,>=3.9.1
#13 4.379   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.414 Collecting filelock<4,>=3.12.2
#13 4.434 Collecting distlib<1,>=0.3.7
#13 4.438   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#13 4.446      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 79.7 MB/s eta 0:00:00
#13 4.446      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 79.7 MB/s eta 0:00:00
#13 4.525 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.714 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.26.6
#13 DONE 4.8s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
    Finished `dev` profile [unoptimized] target(s) in 25.33s
##[endgroup]
error: cannot open .git/FETCH_HEAD: Read-only file system
thread 'main' panicked at /checkout/src/tools/build_helper/src/git.rs:147:54:
called `Result::unwrap()` on an `Err` value: "command did not execute successfully: cd \"/checkout\" && \"git\" \"fetch\" \"origin\" \"master\"\nexpected success, got: exit status: 255\n"
Build completed unsuccessfully in 0:00:35
make: *** [Makefile:98: prepare] Error 1
Command failed. Attempt 2/5:
##[group]Building bootstrap
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
error: cannot open .git/FETCH_HEAD: Read-only file system
thread 'main' panicked at /checkout/src/tools/build_helper/src/git.rs:147:54:
called `Result::unwrap()` on an `Err` value: "command did not execute successfully: cd \"/checkout\" && \"git\" \"fetch\" \"origin\" \"master\"\nexpected success, got: exit status: 255\n"
Build completed unsuccessfully in 0:00:00
make: *** [Makefile:98: prepare] Error 1
Command failed. Attempt 3/5:
##[group]Building bootstrap
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
error: cannot open .git/FETCH_HEAD: Read-only file system
thread 'main' panicked at /checkout/src/tools/build_helper/src/git.rs:147:54:
called `Result::unwrap()` on an `Err` value: "command did not execute successfully: cd \"/checkout\" && \"git\" \"fetch\" \"origin\" \"master\"\nexpected success, got: exit status: 255\n"
Build completed unsuccessfully in 0:00:00
make: *** [Makefile:98: prepare] Error 1
Command failed. Attempt 4/5:
##[group]Building bootstrap
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
error: cannot open .git/FETCH_HEAD: Read-only file system
thread 'main' panicked at /checkout/src/tools/build_helper/src/git.rs:147:54:
called `Result::unwrap()` on an `Err` value: "command did not execute successfully: cd \"/checkout\" && \"git\" \"fetch\" \"origin\" \"master\"\nexpected success, got: exit status: 255\n"
Build completed unsuccessfully in 0:00:00
make: *** [Makefile:98: prepare] Error 1
Command failed. Attempt 5/5:
##[group]Building bootstrap
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
error: cannot open .git/FETCH_HEAD: Read-only file system
thread 'main' panicked at /checkout/src/tools/build_helper/src/git.rs:147:54:
called `Result::unwrap()` on an `Err` value: "command did not execute successfully: cd \"/checkout\" && \"git\" \"fetch\" \"origin\" \"master\"\nexpected success, got: exit status: 255\n"
Build completed unsuccessfully in 0:00:00
make: *** [Makefile:98: prepare] Error 1
The command has failed after 5 attempts.
  local time: Fri Oct 11 14:10:59 UTC 2024
```

</details>

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -122,6 +122,9 @@ jobs:
           # which then uses log commands to actually set them.
           EXTRA_VARIABLES: ${{ toJson(matrix.env) }}
 
+      - name: setup upstream remote
+        run: src/ci/scripts/setup-upstream-remote.sh
+
       - name: ensure the channel matches the target branch
         run: src/ci/scripts/verify-channel.sh
 

--- a/src/ci/scripts/setup-upstream-remote.sh
+++ b/src/ci/scripts/setup-upstream-remote.sh
@@ -1,0 +1,24 @@
+#!/bin/bash
+# In CI environments, bootstrap is forced to use the remote upstream based
+# on "git_repository" and "nightly_branch" values from src/stage0 file.
+# This script configures the remote as it may not exist by default.
+
+set -euo pipefail
+IFS=$'\n\t'
+
+ci_dir=$(cd $(dirname $0) && pwd)/..
+source "$ci_dir/shared.sh"
+
+git_repository=$(parse_stage0_file_by_key "git_repository")
+nightly_branch=$(parse_stage0_file_by_key "nightly_branch")
+
+# Configure "rust-lang/rust" upstream remote only when it's not origin.
+if [ -z "$(git config remote.origin.url | grep $git_repository)" ]; then
+    echo "Configuring https://github.com/$git_repository remote as upstream."
+    git remote add upstream "https://github.com/$git_repository"
+    REMOTE_NAME="upstream"
+else
+    REMOTE_NAME="origin"
+fi
+
+git fetch $REMOTE_NAME $nightly_branch

--- a/src/ci/shared.sh
+++ b/src/ci/shared.sh
@@ -136,3 +136,15 @@ function releaseChannel {
         echo $RUST_CI_OVERRIDE_RELEASE_CHANNEL
     fi
 }
+
+# Parse values from src/stage0 file by key
+function parse_stage0_file_by_key {
+    local key="$1"
+    local file="$ci_dir/../stage0"
+    local value=$(awk -F= '{a[$1]=$2} END {print(a["'$key'"])}' $file)
+    if [ -z "$value" ]; then
+        echo "ERROR: Key '$key' not found in '$file'."
+        exit 1
+    fi
+    echo "$value"
+}

--- a/src/tools/build_helper/src/git.rs
+++ b/src/tools/build_helper/src/git.rs
@@ -1,6 +1,8 @@
 use std::path::{Path, PathBuf};
 use std::process::{Command, Stdio};
 
+use crate::ci::CiEnv;
+
 pub struct GitConfig<'a> {
     pub git_repository: &'a str,
     pub nightly_branch: &'a str,
@@ -114,8 +116,8 @@ fn git_upstream_merge_base(
 
 /// Searches for the nearest merge commit in the repository that also exists upstream.
 ///
-/// If it fails to find the upstream remote, it then looks for the most recent commit made
-/// by the merge bot by matching the author's email address with the merge bot's email.
+/// It looks for the most recent commit made by the merge bot by matching the author's email
+/// address with the merge bot's email.
 pub fn get_closest_merge_commit(
     git_dir: Option<&Path>,
     config: &GitConfig<'_>,
@@ -127,7 +129,15 @@ pub fn get_closest_merge_commit(
         git.current_dir(git_dir);
     }
 
-    let merge_base = git_upstream_merge_base(config, git_dir).unwrap_or_else(|_| "HEAD".into());
+    let merge_base = {
+        if CiEnv::is_ci() {
+            git_upstream_merge_base(config, git_dir).unwrap()
+        } else {
+            // For non-CI environments, ignore rust-lang/rust upstream as it usually gets
+            // outdated very quickly.
+            "HEAD".to_string()
+        }
+    };
 
     git.args([
         "rev-list",

